torch/_higher_order_ops/scan.py

```diff
@@ -181,56 +157,35 @@ def _check_new_carry_match_init(leaves_init, carry_leaves):
     combine_fn = functools.partial(
         wrap_combine_fn_flat,
         combine_fn=combine_fn,
-        spec_init=spec_init,
-        spec_xs=spec_xs,
-        num_init_leaves=len(leaves_init),
-        num_inp_leaves=len(leaves_xs),
+        spec_init=init_spec,
+        spec_xs=xs_spec,
+        num_init_leaves=len(flat_init),
+        num_inp_leaves=len(flat_xs),
     )
 
     def run_flattened_scan(combine_fn, leaves_init, leaves_xs, dim, reverse):
         return scan_op(
-            combine_fn, leaves_init, leaves_xs, dim, reverse, additional_inputs=[]
+            combine_fn, leaves_init, leaves_xs, dim=dim, reverse=reverse, additional_inputs=[]
         )
 
-    if not torch._dynamo.is_compiling():
-        from torch._dynamo.backends.debugging import (
-            make_eager_backend_with_torch_function_mode,
+    carry, result = _maybe_compile_and_run_fn(run_flattened_scan, combine_fn, flat_init, flat_xs, dim, reverse)
+    
+    # TODO: Maybe this check is already too late and should be done earlier?
```

Comment:
It's OK i feel because dynamo cannot see the input spec for carry and we have to do it afterwards.